### PR TITLE
Angular 11 Upgrade 

### PR DESCRIPTION
**Pull request by kdelorey**
### Summary
Upgrade from Angular v7 -> v11. Mostly automated upgrade with a lot of TypeScript and Angular module fixes. We are now using TypeScript ~4.0.

Every warning is fixed and all unit tests are green.

#### Why is the branch called `feature/angular12` but this only upgrades to 11?
Good question, glad you asked! Angular 9+ introduces a new rendering engine that's supposed to be faster at both compile time and runtime, but starting in Angular 12 there is a hard depreciation of the old rendering engine in-favor of the new one (Ivy). I am able to compile using the Ivy render but it seems there are breaking changes in the template generation and the app fails to run at runtime. I didn't spend much time on because of the next point.

Additionally in more recent Angular versions, they switched to using the new `dart-sass` compiler as `node-sass` itself is deprecated. Angular Material 12+ uses a new feature in Sass called modules and it is not implemented in `node-sass` at all. You might be asking, why not just use `dart-sass` then? Unfortunately the sass compile fails with an out of memory exception and I haven't been able get it to work with\ how our our sass sources work right now. 

TL;DR... it won't compile and more work needs to be done in order to get it to work.

### CODE REVIEW

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/screen-parts/membership-display/membership-display.component.html [hunk 1]
```diff
@@ -3,11 +3,6 @@
               [ngClass]="{'in': membership.member, 'not-in': !membership.member}"
               [selectable]="false"
               responsive-class>
-        <app-icon *ngIf="!membership.member && screenData.uncheckMarkIcon"
```
[the remark this answers is not in the document]
Comment (r-cochran) [reply openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/screen-parts/membership-display/membership-display.component.html:6]: Looking at the petco branch, it is extending the class in order to get the X and Check icons injected into the part. It is working on that branch currently. The class it is extending there does the injection constructor it looks like. Confirmed that the X and Check are showing up in 1.12

[the remark this answers is not in the document]
Comment (kdelorey) [reply openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/screen-parts/membership-display/membership-display.component.html:6]: What is the name of the screen part?

[the remark this answers is not in the document]
Comment (r-cochran) [reply openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/screen-parts/membership-display/membership-display.component.html:6]: MembershipDisplayComponentInterface

[the remark this answers is not in the document]
Comment (kdelorey) [reply openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/screen-parts/membership-display/membership-display.component.html]: Okay I just restored it to what it was and fixed the `ctor` so it injected properly and tests pass.

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/lock-screen/lock-screen-data.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { InjectionToken } from "@angular/core";
+import { Observable } from "rxjs";
+import { LockScreenMessage } from "../messages/lock-screen-message";
```
Comment (jchirico8) [inline openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/lock-screen/lock-screen-data.ts:3]: update to single quotes

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/platform-plugins/barcode-scanners/debug-image-scanner/debug-image-scanner.service.ts [hunk 3]
```diff
@@ -1,6 +1,8 @@
 import { Observable } from "rxjs";
 import { ImageScanner, ScanData, ScannerViewRef } from "../scanner";
+import { Injectable } from "@angular/core";
```
Comment (jchirico8) [inline openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/platform-plugins/barcode-scanners/debug-image-scanner/debug-image-scanner.service.ts:3]: Might as well change these to single quotes too

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/keyboard/configs/keyboard-deadkey.config.ts [hunk 4]
```diff
@@ -209,4 +209,4 @@ keyboardDeadkeys['\u00b4'] = keyboardDeadkeys['\u0384'] = keyboardDeadkeys['\'']
 keyboardDeadkeys['\u00ba'] = keyboardDeadkeys['\u00b0'];
 keyboardDeadkeys['\u201a'] = keyboardDeadkeys['\u00B8'];
 
-export { IKeyboardDeadkeys, MAT_KEYBOARD_DEADKEYS, keyboardDeadkeys };
+//export { IKeyboardDeadkeys, MAT_KEYBOARD_DEADKEYS, keyboardDeadkeys };
```
Comment (jchirico8) [inline openpos-client-libs/projects/openpos-client-core-lib/src/lib/keyboard/configs/keyboard-deadkey.config.ts:212]: Can we just delete this export now?

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/screens-with-parts/item-detail/item-detail-option.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import { InjectionToken } from "@angular/core";
```
Comment (jchirico8) [inline openpos-client-libs/projects/openpos-client-core-lib/src/lib/screens-with-parts/item-detail/item-detail-option.ts:1]:
```suggestion
import { InjectionToken } from '@angular/core';
```

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/decorators/screen-part.decorator.ts [hunk 6]
```diff
@@ -1,12 +1,11 @@
+import { Type } from "@angular/core";
```
Comment (jchirico8) [inline openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/decorators/screen-part.decorator.ts:1]:
```suggestion
import { Type } from '@angular/core';
```

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/modal-overlay-container/modal-overlay-container.ts [hunk 7]
```diff
@@ -1,5 +1,7 @@
 import { OverlayContainer } from '@angular/cdk/overlay';
+import { Injectable } from "@angular/core";
```
Comment (jchirico8) [inline openpos-client-libs/projects/openpos-client-core-lib/src/lib/shared/modal-overlay-container/modal-overlay-container.ts:2]:
```suggestion
import { Injectable } from '@angular/core';
```

**Review by jchirico8 — DISMISSED** [review]

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/lock-screen/lock-screen-data.ts (hunk 2 above)
Comment (kdelorey) [reply openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/lock-screen/lock-screen-data.ts:3]: I almost dropped the linter in to enforce things like this, but then I decided I wasn't ready to fixed literally everything... should have just done it 😂

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/keyboard/configs/keyboard-deadkey.config.ts (hunk 4 above)
Comment (kdelorey) [reply openpos-client-libs/projects/openpos-client-core-lib/src/lib/keyboard/configs/keyboard-deadkey.config.ts:212]: Yeah... whoops.

**Review by dankaste — APPROVED** [review]
